### ISSUE TITLE
Contributing to this project

### ISSUE THREAD

**Issue by Errol-Hassall** [ospaarmann/exdgraph#4]
Hi exdgraph
I work for Our Very Own a full service innovation company in Melbourne Australia. Our current project is using Elixir and we decided based on our needs to use a graph database, sepcifically DGraph. There aren't any up to date integrations for DGraph and Elixir except this one. We would like to contribute, we are new to Elixir but are learning fast. Let us know how we can help.

Thanks,
Errol Hassall

**Comment by ospaarmann** [comment]
Hey Errol,

great to hear that! And welcome to Elixir. It is a fantastic language and I'm pretty sure that in 3 years nobody will do ruby anymore...

About your question: I made a couple of design decisions but am otherwise open for input. But first things first before we get fancy.

I generally think that clients for graph databases often make the mistake to try to copy Ecto and its DSL. But this is not fitting for graph dbs since the whole underlying concept is very different. 

**General principles**

- Performance and stability first
- Work as closely to GraphQL+ as possible
- Keep it simple and allow a maximum of flexibility for the user while providing enough syntactic sugar to make it simple to use. Also for people new to Dgraph

**Important things to consider**

- I have started to implement the [DBConnection](https://hexdocs.pm/db_connection/DBConnection.html) behaviour. I will update the Readme shortly. The app already connects on startup to the server using DBConnection and connection pooling (via [Poolboy](https://github.com/devinus/poolboy)).
- Queries are the first thing that I implemented.
- Next would be Mutations etc.
- A good source of inspiration is [Bolt.Sips](https://github.com/florinpatrascu/bolt_sips), an Neo4j driver for Elixir, wrapped around the Bolt protocol. Completely different database and protocol, but uses DBConnection, transactions, connection pooling etc.

**Things that need to be done**

- Test coverage is not great so far. This should be improved.
- One test is failing only on travis. Don't know why.
- Error handling is also not great.
- Check the roadmap in the Readme

**How to proceed**

- Let's agree here what you guys want to work on
- Do a fork and let's go

**Comment by ospaarmann** [comment]
Hey,

a quick update on this:

- Error handling has been improved
- Test coverage is much better
- Travis issue is gone

**Next steps where help is welcome are:**

- Add support for authentication
- Add documentation
- Start with the work on a better query/mutation builder

**Comment by Errol-Hassall** [comment]
Hey

Thanks for getting back so quickly. 

We have actually run into an issue, our setup requires that we use two databases, one being a Postgres db and the other being DGraph, now we have sorted the postgres one out with all the ecto queries we need but when we then add ex_dgrapg to the dependencies it causes conflicts with the postgres db_connection. This can be sorted by updating to the latest and then setting override to true. 
However when you then try and implement a query to dgraph it breaks the postgres connection.

Have you tested using this along side a standard postgres db? Do you have any suggestions on how to proceed.

**Comment by ospaarmann** [comment]
Hey @Errol-Hassall,

I just changed DbConnection to the same version that Ecto uses. Does that fix your issue?

Commit: a81fdf6df670d06535f8c759d7b7a53504cc418d

Cheers

**Comment by Errol-Hassall** [comment]
Hey @ospaarmann 

Yes that worked thanks. 

As for contributing:
As a team we have a combined 2 months experience with Elixir and half of us have the same again for functional programming as a whole. When we feel that we have enough knowledge we will be contributing as we feel strongly about contributing to open source. We will need some time to workout where we can contribute and how.

Thanks,
Errol

**Comment by ospaarmann** [comment]
Great to hear! 

I would suggest: If you are using ExDgraph and find shortcomings, open an issue, discuss and then do a pull request. For example if you need more handy functions for queries or mutations.

In general please open an issue first if you want to do larger changes or introduce new features. Smaller changes or additions to the docs can of course be done directly via a PR.

Thanks

**Comment by Errol-Hassall** [comment]
No worries, will do.

I'm having some trouble running a query. im just running the basic query in the readme. The server is up but it keeps giving me this error.
` ** (EXIT) no process: the process is not alive or there's no process currently associated with the given name, possibly because its application isn't started`
When ever I compile the application. I have added it to the list of applications in mix, but I still get this error.

**Comment by ospaarmann** [comment]
This was on me. I updated the Readme. You also have to add ExDgraph to your suporvisor tree and add configuration.

**Comment by Errol-Hassall** [comment]
Using the map notation which is brilliant I must add. I have entered in a list of people with their attributes. Then after I have created each person I do another iteration linking people to their friends with a friendship edge. What I can't seem to get working is that each person has an id and I forced the uid to be the same so that when I search for someone it makes it easy. 
first iteration:
`map = %{
  uid: person_id
  person_id: person_id,
}`

second iteration:
map = %{
  friendship: %{
  friend: person_id #the other person i want to link to
 }
}
This either doesn't link anything or it just creates a new node, how do I actually link them together with the nodes I created from the first iteration, using the map notation?

**Comment by brianbroderick** [comment]
@ospaarmann If you're still looking for help, I'd be willing to help out.  I have a lot of interest in DGraph, though very little experience with it at the moment. I've been working with Elixir for the past year or so (I've been coding with other languages for a couple of decades now).  I'm starting to learn about DGraph's Go client (I have about a year of Go experience)

**Comment by ospaarmann** [comment]
@brianbroderick great to hear that.

There are two things that need work right now:

- I encountered an issue with Dgraph 1.5. When I use ex_dgraph in a project (umbrella app with a phoenix app inside) the connection to the server keeps breaking every couple of seconds. I described the issue here: https://github.com/tony612/grpc-elixir/issues/61 
- Transactions should be supported and there should be integration tests for them

Help with any of the two would be fantastic.
